### DISCUSSION TITLE
V8 option: --initial-old-space-size

### DISCUSSION THREAD

**Discussion by chenxin1212** [nodejs/help#4776] [Q&A] [answered]
Recently, I have been using Node.js in production and want to try adjusting V8 startup parameters like JVM to improve the throughput of the service. When testing the parameter ```--initial-old-space-size``` I found that configuring this parameter did not have any impact on heap and garbage collection. Has anyone encountered a similar issue? Is this a bug or is there some other logic at play?

**Comment by chenxin1212** [comment]
node version: 18.12.1

**Comment by bnoordhuis** [comment]
You probably set it too low. It defaults to 256 or 512 MB (32 vs. 64 bits) and you can't set it higher than half the max old space size (so 350 or 700 MB unless overridden with `--max-old-space-size=..`.)

I don't expect higher values to make much difference with most workloads. You avoid a few full GCs at startup but that's about it. Possibly useful for short-lived scripts but not long-lived services.

**Reply by chenxin1212** [reply]
Thank you for your reply.

 I tried to read the code in heap.cc, but... it's quite extensive and I haven't fully understood it yet. Regarding the mentioned initial value or the value set by "--initial-old-space-size", can it be confirmed through "v8.getHeapSpaceSizeStatistics" API? Currently, I'm using this API to confirm, but found inconsistencies.

 Besides trying to adjust the parameters to optimize the service, I also want to explore more about GC knowledge through this. Looking forward to your response, thank you.

**Reply by bnoordhuis** [reply, marked as answer]
I don't think you can observe it directly. At best the physical_space_size value grows a little quicker than it otherwise would.

It's probably best to think of it as an input signal (but not the only one - it's complicated) that helps V8 decide how quickly to grow the heap to the max size.

**Reply by chenxin1212** [reply]
Got it, thank you so much. I thoroughly tested it in the live service and the results perfectly align with your description. At this point, I won't explicitly define these parameters in the service. Instead, I'll assess them based on online traffic conditions and my understanding of GC expertise.
